# Data Science Academy

## Fundamentals of Python Language for Data Analyse and Data Science

### Manipulation of Data with Pandas

In [1]:
# Python Language version
from platform import python_version
print("Python Language version used in this Jupyter Notebook: ", python_version())

Python Language version used in this Jupyter Notebook:  3.11.5


### User Guide Pandas:

https://pandas.pydata.org/docs/user_guide/index.html

In [2]:
# Install the exact package version (if necessary)
!pip install -q pandas==1.5.3

In [3]:
import pandas as pd

In [4]:
pd.__version__

'2.0.3'

### Manipulating Data in DataFrames from Pandas

In [5]:
# Create a dictionary
data = {"State": ["Santa Catarina", "Rio de Janeiro", "Pernambuco", "Bahia", "Minas Gerais"], 
        "Year": [2004, 2005, 2006, 2007, 2008],
        "Employement Rate": [1.5, 1.7, 1.6, 2.4, 2.7]}

In [6]:
# Import the function DataFrame from Pandas
from pandas import DataFrame

In [7]:
# Convert the dicitonary in a dataframe
df = DataFrame(data)

In [8]:
# Visualize the first 5 lines
df.head()

State  Year  Employement Rate
0  Santa Catarina  2004               1.5
1  Rio de Janeiro  2005               1.7
2      Pernambuco  2006               1.6
3           Bahia  2007               2.4
4    Minas Gerais  2008               2.7

In [9]:
type(df)

pandas.core.frame.DataFrame

In [10]:
# Reorganizing the columns
DataFrame(data, columns = ["State", "Employement Rate", "Year"])

State  Employement Rate  Year
0  Santa Catarina               1.5  2004
1  Rio de Janeiro               1.7  2005
2      Pernambuco               1.6  2006
3           Bahia               2.4  2007
4    Minas Gerais               2.7  2008

In [11]:
df2 = DataFrame(data, columns=["State", "Employement Rate", "Grow Rate", "Year",],
               index = ["state1", "state2", "state3", "state4", "state5"])

In [12]:
df2

State  Employement Rate Grow Rate  Year
state1  Santa Catarina               1.5       NaN  2004
state2  Rio de Janeiro               1.7       NaN  2005
state3      Pernambuco               1.6       NaN  2006
state4           Bahia               2.4       NaN  2007
state5    Minas Gerais               2.7       NaN  2008

In [13]:
df2.columns

Index(['State', 'Employement Rate', 'Grow Rate', 'Year'], dtype='object')

In [14]:
# Printing only one column from the DataFrame
df2["State"]

state1    Santa Catarina
state2    Rio de Janeiro
state3        Pernambuco
state4             Bahia
state5      Minas Gerais
Name: State, dtype: object

In [15]:
# Printing only two columns from the DataFrame
df2[["Employement Rate", "Year"]]

Employement Rate  Year
state1               1.5  2004
state2               1.7  2005
state3               1.6  2006
state4               2.4  2007
state5               2.7  2008

In [16]:
df2.index

Index(['state1', 'state2', 'state3', 'state4', 'state5'], dtype='object')

In [17]:
df2.values

array([['Santa Catarina', 1.5, nan, 2004],
       ['Rio de Janeiro', 1.7, nan, 2005],
       ['Pernambuco', 1.6, nan, 2006],
       ['Bahia', 2.4, nan, 2007],
       ['Minas Gerais', 2.7, nan, 2008]], dtype=object)

In [18]:
df2.dtypes

State                object
Employement Rate    float64
Grow Rate            object
Year                  int64
dtype: object

In [19]:
# Python Language is case sensitive
# df2["state"]

In [20]:
df2.filter(items = ["state3"], axis = 0)

State  Employement Rate Grow Rate  Year
state3  Pernambuco               1.6       NaN  2006

### Using NumPy and Pandas for Data Manipulation

In [21]:
# Importing Numpy
import numpy as np

In [22]:
df2.head()

State  Employement Rate Grow Rate  Year
state1  Santa Catarina               1.5       NaN  2004
state2  Rio de Janeiro               1.7       NaN  2005
state3      Pernambuco               1.6       NaN  2006
state4           Bahia               2.4       NaN  2007
state5    Minas Gerais               2.7       NaN  2008

In [26]:
df2.describe()

Employement Rate         Year
count          5.000000     5.000000
mean           1.980000  2006.000000
std            0.535724     1.581139
min            1.500000  2004.000000
25%            1.600000  2005.000000
50%            1.700000  2006.000000
75%            2.400000  2007.000000
max            2.700000  2008.000000

In [23]:
df2.isna()

State  Employement Rate  Grow Rate   Year
state1  False             False       True  False
state2  False             False       True  False
state3  False             False       True  False
state4  False             False       True  False
state5  False             False       True  False

In [24]:
df2["Grow Rate"].isna()

state1    True
state2    True
state3    True
state4    True
state5    True
Name: Grow Rate, dtype: bool

In [27]:
df2["Grow Rate"] = np.arange(5.)

In [28]:
df2

State  Employement Rate  Grow Rate  Year
state1  Santa Catarina               1.5        0.0  2004
state2  Rio de Janeiro               1.7        1.0  2005
state3      Pernambuco               1.6        2.0  2006
state4           Bahia               2.4        3.0  2007
state5    Minas Gerais               2.7        4.0  2008

In [29]:
df2.dtypes

State                object
Employement Rate    float64
Grow Rate           float64
Year                  int64
dtype: object

In [30]:
df2["Grow Rate"].isna()

state1    False
state2    False
state3    False
state4    False
state5    False
Name: Grow Rate, dtype: bool

In [31]:
# Statistic resume of Dataframe
df2.describe()

Employement Rate  Grow Rate         Year
count          5.000000   5.000000     5.000000
mean           1.980000   2.000000  2006.000000
std            0.535724   1.581139     1.581139
min            1.500000   0.000000  2004.000000
25%            1.600000   1.000000  2005.000000
50%            1.700000   2.000000  2006.000000
75%            2.400000   3.000000  2007.000000
max            2.700000   4.000000  2008.000000

### Slicing Pandas Dataframes

In [32]:
df2

State  Employement Rate  Grow Rate  Year
state1  Santa Catarina               1.5        0.0  2004
state2  Rio de Janeiro               1.7        1.0  2005
state3      Pernambuco               1.6        2.0  2006
state4           Bahia               2.4        3.0  2007
state5    Minas Gerais               2.7        4.0  2008

In [37]:
df2["state2": "state4"]

State  Employement Rate  Grow Rate  Year
state2  Rio de Janeiro               1.7        1.0  2005
state3      Pernambuco               1.6        2.0  2006
state4           Bahia               2.4        3.0  2007

In [35]:
df2[df2["Employement Rate"] < 2]

State  Employement Rate  Grow Rate  Year
state1  Santa Catarina               1.5        0.0  2004
state2  Rio de Janeiro               1.7        1.0  2005
state3      Pernambuco               1.6        2.0  2006

In [36]:
df2[["State", "Grow Rate"]]

State  Grow Rate
state1  Santa Catarina        0.0
state2  Rio de Janeiro        1.0
state3      Pernambuco        2.0
state4           Bahia        3.0
state5    Minas Gerais        4.0

In [38]:
df2[["State", "Grow Rate", "Employement Rate"]]

State  Grow Rate  Employement Rate
state1  Santa Catarina        0.0               1.5
state2  Rio de Janeiro        1.0               1.7
state3      Pernambuco        2.0               1.6
state4           Bahia        3.0               2.4
state5    Minas Gerais        4.0               2.7

### Filling Absent Values in Pandas DataFrames

The function fillna() is used to fill absent values. The function offers many options. We can use a specific value, an aggregate functon (for instance, average() or the previous values or the next.

For that example we'll use the mode, the statistic represents the value that appear more times in a variable

In [40]:
# First we import a dataset
df = pd.read_csv("dataset.csv")

In [45]:
df.head(5) # 5 is the default value

ID_Pedido Data_Pedido ID_Cliente   Segmento           Pais Regiao  \
0  CA-2016-152156  2016-11-08   CG-12520   Consumer  United States  South   
1  CA-2016-152156  2016-11-08   CG-12520   Consumer  United States  South   
2  CA-2016-138688  2016-06-12   DV-13045  Corporate  United States   West   
3  US-2015-108966  2015-10-11   SO-20335   Consumer  United States  South   
4  US-2015-108966  2015-10-11   SO-20335   Consumer  United States  South   

        ID_Produto        Categoria  \
0  FUR-BO-10001798        Furniture   
1  FUR-CH-10000454        Furniture   
2  OFF-LA-10000240  Office Supplies   
3  FUR-TA-10000577        Furniture   
4  OFF-ST-10000760  Office Supplies   

                                        Nome_Produto  Valor_Venda  Quantidade  
0                  Bush Somerset Collection Bookcase     261.9600         NaN  
1  Hon Deluxe Fabric Upholstered Stacking Chairs,...     731.9400         NaN  
2  Self-Adhesive Address Labels for Typewriters b...      14.6200         2.0  
3      Bretford CR4500 Series Slim Rectangular Table     957.5775         5.0  
4                     Eldon Fold 'N Roll Cart System      22.3680         2.0

In [42]:
df.isna().sum()

ID_Pedido       0
Data_Pedido     0
ID_Cliente      0
Segmento        0
Pais            0
Regiao          0
ID_Produto      0
Categoria       0
Nome_Produto    0
Valor_Venda     0
Quantidade      2
dtype: int64

In [44]:
# We extract the mode of Quantity column
mode = df["Quantidade"].value_counts().index[0]

In mathematics, the mode is the value that appears most frequently in a set of data. It's a measure of central tendency, like the mean and median, used to describe the typical value in a dataset. 

In data analysis, the mode helps identify the most common occurrence or trend within a dataset, which can be useful for various applications. 

In [47]:
print(mode)

3.0


In [50]:
# And in the end we fill the values NA with the mode
df["Quantidade"].fillna(value = mode, inplace = True) 
# inplace = True, it tells pandas to replace the values in the dataframe
# itself instead of creating a copy

In [49]:
df.isna().sum()

ID_Pedido       0
Data_Pedido     0
ID_Cliente      0
Segmento        0
Pais            0
Regiao          0
ID_Produto      0
Categoria       0
Nome_Produto    0
Valor_Venda     0
Quantidade      0
dtype: int64

### Data Query in Pandas DataFrame

With the pandas, we can create data frames. which are essentially tables. In such a way, we can make queries. And for that, we use the consult method query(). See the example below:

In [51]:
# We check the minimum and max values of the colum Valor_Venda
df.Valor_Venda.describe()

count     9994.000000
mean       229.858001
std        623.245101
min          0.444000
25%         17.280000
50%         54.490000
75%        209.940000
max      22638.480000
Name: Valor_Venda, dtype: float64

The sales interval (Valor_Venda) is of 0.44 to 22638. Let's make a consultation and return all the sales inside a range of values. We make this as the instruction below:

In [52]:
# We generate a new dataframe only with the sales interval 
# between 229 and 10000
df2 = df.query("229 < Valor_Venda < 10000")

In [53]:
# Then, we confirm the minimum and max values
df2.Valor_Venda.describe()

count    2357.000000
mean      766.679142
std       856.315136
min       229.544000
25%       323.100000
50%       490.320000
75%       859.200000
max      9892.740000
Name: Valor_Venda, dtype: float64

In [54]:
# We generate a new dataframe only with the sale values above the average
df3 = df2.query("Valor_Venda > 766")

In [55]:
df3.head()

ID_Pedido Data_Pedido ID_Cliente  Segmento           Pais Regiao  \
3   US-2015-108966  2015-10-11   SO-20335  Consumer  United States  South   
7   CA-2014-115812  2014-06-09   BH-11710  Consumer  United States   West   
10  CA-2014-115812  2014-06-09   BH-11710  Consumer  United States   West   
11  CA-2014-115812  2014-06-09   BH-11710  Consumer  United States   West   
24  CA-2015-106320  2015-09-25   EB-13870  Consumer  United States   West   

         ID_Produto   Categoria  \
3   FUR-TA-10000577   Furniture   
7   TEC-PH-10002275  Technology   
10  FUR-TA-10001539   Furniture   
11  TEC-PH-10002033  Technology   
24  FUR-TA-10000577   Furniture   

                                     Nome_Produto  Valor_Venda  Quantidade  
3   Bretford CR4500 Series Slim Rectangular Table     957.5775         5.0  
7                  Mitel 5320 IP Phone VoIP phone     907.1520         6.0  
10       Chromcraft Rectangular Conference Tables    1706.1840         9.0  
11  Konftel 250 Conference phone - Charcoal black     911.4240         4.0  
24  Bretford CR4500 Series Slim Rectangular Table    1044.6300         3.0

Query executed succesfully!!

### Checking the Occurrence of Several Values in a Column

In our example dataset we have the column Quantidade that represents the quantity of sold items in each one of the sales. Imagine that we need to know in which sales were sold 5, 7, 9 or 11 items.

Will we apply that type of filter in our dataframe?

Easy. Pandas offers the method isin() to check several values in a column. Who knows SQL Language must already have noticed that the method is the same as the clause IN in SQL. Let's go to the example.

In [58]:
df.shape

(9994, 11)

In [59]:
# Then we apply the filter
df[df["Quantidade"].isin([5, 7, 9, 11])]

ID_Pedido Data_Pedido ID_Cliente     Segmento           Pais  \
3     US-2015-108966  2015-10-11   SO-20335     Consumer  United States   
5     CA-2014-115812  2014-06-09   BH-11710     Consumer  United States   
9     CA-2014-115812  2014-06-09   BH-11710     Consumer  United States   
10    CA-2014-115812  2014-06-09   BH-11710     Consumer  United States   
14    US-2015-118983  2015-11-22   HP-14815  Home Office  United States   
...              ...         ...        ...          ...            ...   
9974  US-2016-103674  2016-12-06   AP-10720  Home Office  United States   
9977  US-2016-103674  2016-12-06   AP-10720  Home Office  United States   
9981  CA-2017-163566  2017-08-03   TB-21055     Consumer  United States   
9982  US-2016-157728  2016-09-22   RC-19960     Consumer  United States   
9988  CA-2017-163629  2017-11-17   RA-19885    Corporate  United States   

       Regiao       ID_Produto        Categoria  \
3       South  FUR-TA-10000577        Furniture   
5        West  FUR-FU-10001487        Furniture   
9        West  OFF-AP-10002892  Office Supplies   
10       West  FUR-TA-10001539        Furniture   
14    Central  OFF-AP-10002311  Office Supplies   
...       ...              ...              ...   
9974     West  OFF-AR-10004752  Office Supplies   
9977     West  OFF-FA-10003467  Office Supplies   
9981     East  OFF-LA-10004484  Office Supplies   
9982  Central  OFF-PA-10002195  Office Supplies   
9988    South  TEC-PH-10004006       Technology   

                                           Nome_Produto  Valor_Venda  \
3         Bretford CR4500 Series Slim Rectangular Table     957.5775   
5     Eldon Expressions Wood and Plastic Desk Access...      48.8600   
9                      Belkin F5C206VTEL 6 Outlet Surge     114.9000   
10             Chromcraft Rectangular Conference Tables    1706.1840   
14    Holmes Replacement Filter for HEPA Air Cleaner...      68.8100   
...                                                 ...          ...   
9974                               Blackstonian Pencils      18.6900   
9977           Alliance Big Bands Rubber Bands, 12/Pack      13.8600   
9981                                          Avery 476      16.5200   
9982  RSVP Cards & Envelopes, Blank White, 8-1/2" X ...      35.5600   
9988                    Panasonic KX - TS880B Telephone     206.1000   

      Quantidade  
3            5.0  
5            7.0  
9            5.0  
10           9.0  
14           5.0  
...          ...  
9974         7.0  
9977         7.0  
9981         5.0  
9982         7.0  
9988         5.0  

[2128 rows x 11 columns]

In the instruction above, we're filtering the dataframe called f, returning all the lines where the column Quantity is equal to the values 5, 7, 9 or 11. We give a list of values as argument for the method isin().

Let's make it a little more fun. If you executed the instruction above, you noticed that were returned 2.128 lines. And if we want to return only 10 lines? Just slice the result like this:

In [61]:
# Shape
df[df["Quantidade"].isin([5, 7, 9, 11])].shape

(2128, 11)

In [63]:
df[df["Quantidade"].isin([5, 7, 9, 11])][:10]

ID_Pedido Data_Pedido ID_Cliente     Segmento           Pais  \
3   US-2015-108966  2015-10-11   SO-20335     Consumer  United States   
5   CA-2014-115812  2014-06-09   BH-11710     Consumer  United States   
9   CA-2014-115812  2014-06-09   BH-11710     Consumer  United States   
10  CA-2014-115812  2014-06-09   BH-11710     Consumer  United States   
14  US-2015-118983  2015-11-22   HP-14815  Home Office  United States   
21  CA-2016-137330  2016-12-09   KB-16585    Corporate  United States   
22  CA-2016-137330  2016-12-09   KB-16585    Corporate  United States   
27  US-2015-150630  2015-09-17   TB-21520     Consumer  United States   
35  CA-2016-117590  2016-12-08   GH-14485    Corporate  United States   
36  CA-2016-117590  2016-12-08   GH-14485    Corporate  United States   

     Regiao       ID_Produto        Categoria  \
3     South  FUR-TA-10000577        Furniture   
5      West  FUR-FU-10001487        Furniture   
9      West  OFF-AP-10002892  Office Supplies   
10     West  FUR-TA-10001539        Furniture   
14  Central  OFF-AP-10002311  Office Supplies   
21  Central  OFF-AR-10000246  Office Supplies   
22  Central  OFF-AP-10001492  Office Supplies   
27     East  FUR-BO-10004834        Furniture   
35  Central  TEC-PH-10004977       Technology   
36  Central  FUR-FU-10003664        Furniture   

                                         Nome_Produto  Valor_Venda  Quantidade  
3       Bretford CR4500 Series Slim Rectangular Table     957.5775         5.0  
5   Eldon Expressions Wood and Plastic Desk Access...      48.8600         7.0  
9                    Belkin F5C206VTEL 6 Outlet Surge     114.9000         5.0  
10           Chromcraft Rectangular Conference Tables    1706.1840         9.0  
14  Holmes Replacement Filter for HEPA Air Cleaner...      68.8100         5.0  
21                                         Newell 318      19.4600         7.0  
22        Acco Six-Outlet Power Strip, 4' Cord Length      60.3400         7.0  
27  Riverside Palais Royal Lawyers Bookcase, Royal...    3083.4300         7.0  
35                                        GE 30524EE4    1097.5440         7.0  
36  Electrix Architect's Clamp-On Swing Arm Lamp, ...     190.9200         5.0

Isn't Pandas incredible?

### Logical Operators for Data Manipulation with Pandas

The Logical Operators are excelent to filter dataframes and return exactly the data we need for our work. To know more about the logical operators rules, go to this page:

https://press.rebus.community/programmingfundamentals/chapter/logical-operators/

## End
### Thank you!